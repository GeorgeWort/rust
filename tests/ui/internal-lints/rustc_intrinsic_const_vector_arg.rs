// compile-flags: -Z unstable-options
#![feature(abi_unadjusted)]
#![feature(inline_const)]
#![feature(intrinsics)]
#![allow(non_camel_case_types)]
#![feature(repr_simd)]
#![feature(rustc_attrs)]
#![feature(simd_ffi)]
#![allow(unused)]

#[repr(simd)]
#[derive(Clone)]
pub struct i8x2(i8, i8);

extern "unadjusted" {
    #[rustc_intrinsic_const_vector_arg] //~ ERROR malformed `rustc_intrinsic_const_vector_arg` attribute input
    fn foo1(a: i8x2, b: i8);
}

extern "unadjusted" {
    #[rustc_intrinsic_const_vector_arg = "1"] //~ ERROR malformed `rustc_intrinsic_const_vector_arg` attribute input
    fn foo2(a: i8x2, b: i8);
}

#[rustc_intrinsic_const_vector_arg(0)] //~ ERROR  attribute should be applied to functions in `extern "unadjusted"` modules
pub struct foo3(i8x2);

extern "C" {
    #[rustc_intrinsic_const_vector_arg(0)] //~ ERROR  attribute should be applied to functions in `extern "unadjusted"` modules
    fn foo4(a: i8x2);
}

extern "unadjusted" {
    #[rustc_intrinsic_const_vector_arg(0)] //~ ERROR function does not have a parameter at index 0
    fn foo5();
}

extern "unadjusted" {
    #[rustc_intrinsic_const_vector_arg(1)] //~ ERROR function does not have a parameter at index 1
    fn foo6(a: i8x2);
}

extern "unadjusted" {
    #[rustc_intrinsic_const_vector_arg("bar")] //~ ERROR attribute requires a parameter index
    fn foo7(a: i8x2);
}

extern "unadjusted" {
    #[rustc_intrinsic_const_vector_arg(0,2)] //~ ERROR function does not have a parameter at index 2
    fn foo8(a: i8x2, b: i8);
}

extern "unadjusted" {
    #[rustc_intrinsic_const_vector_arg(0)] //~ ERROR parameter at index 0 must be a simd type
    fn foo9(a: i8);
}

fn main() {}
