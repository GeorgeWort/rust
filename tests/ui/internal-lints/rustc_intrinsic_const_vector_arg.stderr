error: malformed `rustc_intrinsic_const_vector_arg` attribute input
  --> $DIR/rustc_intrinsic_const_vector_arg.rs:16:5
   |
LL |     #[rustc_intrinsic_const_vector_arg]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: must be of the form: `#[rustc_intrinsic_const_vector_arg(arg1_index, arg2_index, ...)]`

error: malformed `rustc_intrinsic_const_vector_arg` attribute input
  --> $DIR/rustc_intrinsic_const_vector_arg.rs:21:5
   |
LL |     #[rustc_intrinsic_const_vector_arg = "1"]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: must be of the form: `#[rustc_intrinsic_const_vector_arg(arg1_index, arg2_index, ...)]`

error: attribute should be applied to functions in `extern "unadjusted"` modules
  --> $DIR/rustc_intrinsic_const_vector_arg.rs:25:1
   |
LL | #[rustc_intrinsic_const_vector_arg(0)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
LL | pub struct foo3(i8x2);
   | ---------------------- not a function in an `extern "unadjusted"` module

error: attribute should be applied to functions in `extern "unadjusted"` modules
  --> $DIR/rustc_intrinsic_const_vector_arg.rs:29:5
   |
LL |     #[rustc_intrinsic_const_vector_arg(0)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
LL |     fn foo4(a: i8x2);
   |     ----------------- not a function in an `extern "unadjusted"` module

error: function does not have a parameter at index 0
  --> $DIR/rustc_intrinsic_const_vector_arg.rs:34:5
   |
LL |     #[rustc_intrinsic_const_vector_arg(0)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
LL |     fn foo5();
   |     ---------- function has 0 arguments

error: function does not have a parameter at index 1
  --> $DIR/rustc_intrinsic_const_vector_arg.rs:39:5
   |
LL |     #[rustc_intrinsic_const_vector_arg(1)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
LL |     fn foo6(a: i8x2);
   |     ----------------- function has 1 arguments

error: attribute requires a parameter index
  --> $DIR/rustc_intrinsic_const_vector_arg.rs:44:40
   |
LL |     #[rustc_intrinsic_const_vector_arg("bar")]
   |                                        ^^^^^

error: function does not have a parameter at index 2
  --> $DIR/rustc_intrinsic_const_vector_arg.rs:49:5
   |
LL |     #[rustc_intrinsic_const_vector_arg(0,2)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
LL |     fn foo8(a: i8x2, b: i8);
   |     ------------------------ function has 2 arguments

error: parameter at index 0 must be a simd type
  --> $DIR/rustc_intrinsic_const_vector_arg.rs:54:5
   |
LL |     #[rustc_intrinsic_const_vector_arg(0)]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
LL |     fn foo9(a: i8);
   |                -- parameter is a non-simd type

error: aborting due to 9 previous errors

